File: ruby/lib/scc.rb
# 強連結成分(Strongly Connected Components)

class SCC
  attr_reader :size, :edges
  def initialize(n)
    @size = n
    @edges = Array.new(size) { [] }
  end

  def add_edge(from, to)
    edges[from] << to
  end

  def add_edges(arr)
    arr.each { add_edge(_1, _2) }
  end

  def scc
    group_num, ids = scc_ids
    groups = Array.new(group_num) { [] }
    ids.each_with_index { groups[_1] << _2 }
    groups
  end

  private
  def scc_ids
    now_ord = group_num = 0
    visited = []
    low = []
    ord = []
    ids = []

    dfs =-> from do
      low[from] = ord[from] = now_ord
      now_ord += 1
      visited << from

      edges[from].each do |to|
        if ord[to]
          low[from] = [low[from], ord[to]].min
        else
          dfs.call(to)
          low[from] = [low[from], low[to]].min
        end
      end
      return unless low[from] == ord[from]
      while (to = visited.pop)
        # back edgeやcross edgeはordが小さい頂点と結ばれることがあるため、
        # 頂点の SCC id が確定している場合にも頂点の lowlink が更新されてしまう可能性があるため
        # ordをlowlinkの更新に影響しない値に置き換えておく。
        ord[to] = size
        ids[to] = group_num
        break if from == to
      end
      group_num += 1
    end

    size.times do |i|
      next if ord[i]
      dfs.call(i)
    end

    ids.map! { group_num - _1 - 1 }

    [group_num, ids]
  end
end

File: javascript/lib/scc.js
// 強連結成分(Strongly Connected Components)
// 無向グラフでは使用できない

class SCC {
  constructor(size) {
    this.size = size;
    this.edges = [...Array(this.size)].map(() => []);
  }

  addEdge(from, to) {
    this.edges[from].push(to);
  }

  addEdges(edges) {
    edges.forEach((edge) => this.addEdge(...edge));
  }

  scc() {
    const [groupSize, ids] = this.#sccIds();
    const groups = [...Array(groupSize)].map(() => []);
    ids.forEach((id, i) => groups[id].push(i));
    return groups;
  }

  #sccIds() {
    let groupNum = 0;
    let currentOrd = 0;
    const visited = [];
    const ord = [];
    const low = [];
    const ids = [];

    const dfs = from => {
      if (ord[from]) return;
      ord[from] = currentOrd;
      low[from] = currentOrd++;
      visited.push(from);

      this.edges[from].forEach(to => {
        if (ord[to]) {
          low[from] = Math.min(low[from], ord[to]);
        } else {
          dfs(to);
          low[from] = Math.min(low[from], low[to]);
        }
      });

      if (ord[from] !== low[from]) return;
      while (visited.length !== 0) {
        const to = visited.pop();
        ord[to] = this.size;
        ids[to] = groupNum;
        if (from == to) break;
      }
      groupNum++;
    }

    for (let i = 0; i < this.size; i++) dfs(i);

    return [groupNum, ids.map(i => groupNum - 1 - i)];
  }
}

module.exports = SCC;
